src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretManagerProvider.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    public sealed class DefaultSecretManagerProvider : ISecretManagerProvider
+    {
+        private readonly ILogger _logger;
+        private readonly IOptionsMonitor<ScriptApplicationHostOptions> _options;
+        private readonly IHostIdProvider _hostIdProvider;
+        private readonly IConfiguration _configuration;
+        private readonly IEnvironment _environment;
+        private Lazy<ISecretManager> _secretManagerLazy;
+
+        public DefaultSecretManagerProvider(IOptionsMonitor<ScriptApplicationHostOptions> options, IHostIdProvider hostIdProvider,
+            IConfiguration configuration, IEnvironment environment, ILoggerFactory loggerFactory)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _hostIdProvider = hostIdProvider ?? throw new ArgumentNullException(nameof(hostIdProvider));
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+
+            _logger = loggerFactory.CreateLogger(ScriptConstants.LogCategoryHostGeneral);
+            _secretManagerLazy = new Lazy<ISecretManager>(Create);
+
+            // When these options change (due to specialization), we need to reset the secret manager.
+            options.OnChange(_ => ResetSecretManager());
```

Comment:
Because the SecretManager runs at the WebHost in standby mode, we need to make sure we refresh secrets during specialization. 

In reality, the SecretManager likely will never be loaded in standby mode, but this gives us some guarantees.